cooked-validators/src/Cooked/MockChain/BlockChain.hs

```diff
@@ -147,78 +196,99 @@ datumFromTxOutRef oref = do
-  allUtxos = lift allUtxos
-  txOutByRef = lift . txOutByRef
+instance (MonadTransControl t, MonadError MockChainError m, Monad (t m)) => MonadError MockChainError (AsTrans t m) where
+  throwError = lift . throwError
+  catchError act f = liftWith (\run -> catchError (run act) (run . f)) >>= restoreT . return
+
+instance (MonadTrans t, MonadBlockChainBalancing m, Monad (t m), MonadError MockChainError (AsTrans t m)) => MonadBlockChainBalancing (AsTrans t m) where
+  getParams = lift getParams
+  validatorFromHash = lift . validatorFromHash
+  utxosAtLedger = lift . utxosAtLedger
+  txOutByRefLedger = lift . txOutByRefLedger
   datumFromHash = lift . datumFromHash
+
+instance (MonadTrans t, MonadBlockChainWithoutValidation m, Monad (t m), MonadError MockChainError (AsTrans t m)) => MonadBlockChainWithoutValidation (AsTrans t m) where
+  allUtxosLedger = lift allUtxosLedger
   ownPaymentPubKeyHash = lift ownPaymentPubKeyHash
   currentSlot = lift currentSlot
   currentTime = lift currentTime
   awaitSlot = lift . awaitSlot
   awaitTime = lift . awaitTime
 
-instance (MonadTrans t, MonadBlockChain m, MonadFail (t m)) => MonadBlockChain (AsTrans t m) where
+instance (MonadTrans t, MonadBlockChain m, MonadBlockChainWithoutValidation (AsTrans t m)) => MonadBlockChain (AsTrans t m) where
   validateTxSkel = lift . validateTxSkel
 
+deriving via (AsTrans (WriterT w) m) instance (Monoid w, MonadBlockChainBalancing m) => MonadBlockChainBalancing (WriterT w m)
+
 deriving via (AsTrans (WriterT w) m) instance (Monoid w, MonadBlockChainWithoutValidation m) => MonadBlockChainWithoutValidation (WriterT w m)
 
 deriving via (AsTrans (WriterT w) m) instance (Monoid w, MonadBlockChain m) => MonadBlockChain (WriterT w m)
 
+deriving via (AsTrans (ReaderT r) m) instance MonadBlockChainBalancing m => MonadBlockChainBalancing (ReaderT r m)
+
 deriving via (AsTrans (ReaderT r) m) instance MonadBlockChainWithoutValidation m => MonadBlockChainWithoutValidation (ReaderT r m)
 
 deriving via (AsTrans (ReaderT r) m) instance MonadBlockChain m => MonadBlockChain (ReaderT r m)
 
+deriving via (AsTrans (StateT s) m) instance MonadBlockChainBalancing m => MonadBlockChainBalancing (StateT s m)
+
 deriving via (AsTrans (StateT s) m) instance MonadBlockChainWithoutValidation m => MonadBlockChainWithoutValidation (StateT s m)
 
 deriving via (AsTrans (StateT s) m) instance MonadBlockChain m => MonadBlockChain (StateT s m)
 
-deriving via (AsTrans ListT m) instance MonadBlockChainWithoutValidation m => MonadBlockChainWithoutValidation (ListT m)
+-- 'ListT' has no 'MonadTransControl' instance, so the @deriving via ...@
+-- machinery is unusable here. However, there is
+--
+-- > MonadError e m => MonadError e (ListT m)
+--
+-- so I decided to go with a bit of code duplication to implement the
+-- 'MonadBlockChainWithoutValidation' and 'MonadBlockChain' instances for
+-- 'ListT', instead of more black magic...
```

Comment:
It implements `MonadBaseControl` though, so I'm curious if I can hack around this later on.

Not a big deal right now, of course!